spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/reactor/ReactorSleuth.java

```diff
@@ -59,65 +57,69 @@ private ReactorSleuth() {
+			// excessive overhead when using Flux/Mono #just, #empty, #error, etc.
 			if (p instanceof Fuseable.ScalarCallable) {
 				return sub;
 			}
 
-			if (beanFactory instanceof ConfigurableApplicationContext
-					&& ((ConfigurableApplicationContext) beanFactory).isActive()) {
-				if (log.isTraceEnabled()) {
-					log.trace("Spring Context [" + beanFactory
-							+ "] already refreshed. Creating a scope "
-							+ "passing span subscriber with Reactor Context " + "["
-							+ sub.currentContext() + "] and name [" + name(sub) + "]");
+			Context context = sub.currentContext();
+
+			if (log.isTraceEnabled()) {
+				log.trace("Spring context [" + beanFactory + "], Reactor context ["
+						+ context + "], name [" + name(sub) + "]");
+			}
+
+			// Check fast path, where we don't need to defer access to the trace context
+			CurrentTraceContext unwrapped = lazyCurrentTraceContext.unwrap();
+			if (unwrapped != null) {
+				TraceContext parent = traceContext(context, unwrapped);
+				if (parent == null) {
+					return sub; // no need to trace
 				}
 
-				return scopePassingSpanSubscription(beanFactory, sub);
+				if (log.isTraceEnabled()) {
+					log.trace(
+							"Creating a scope passing span subscriber with Reactor Context "
+									+ "[" + context + "] and name [" + name(sub) + "]");
+				}
+				return new ScopePassingSpanSubscriber<>(sub, context, unwrapped, parent);
 			}
 			if (log.isTraceEnabled()) {
-				log.trace("Spring Context [" + beanFactory
-						+ "] is not yet refreshed, falling back to lazy span subscriber. Reactor Context is ["
+				log.trace("Falling back to lazy span subscriber. Reactor Context is ["
 						+ sub.currentContext() + "] and name is [" + name(sub) + "]");
 			}
-			return new LazySpanSubscriber<>(
-					new SpanSubscriptionProvider<>(beanFactory, sub));
+			return new LazyScopePassingSpanSubscriber<>(sub, context,
+					lazyCurrentTraceContext);
 		});
 	}
 
 	static String name(CoreSubscriber<?> sub) {
 		return Scannable.from(sub).name();
 	}
 
-	private static Map<BeanFactory, CurrentTraceContext> CACHE = new ConcurrentHashMap<>();
-
-	static <T> CoreSubscriber<? super T> scopePassingSpanSubscription(
-			BeanFactory beanFactory, CoreSubscriber<? super T> sub) {
-		CurrentTraceContext currentTraceContext = CACHE.computeIfAbsent(beanFactory,
-				beanFactory1 -> beanFactory1.getBean(CurrentTraceContext.class));
-		Context context = sub.currentContext();
-
-		TraceContext parent = context.getOrDefault(TraceContext.class, null);
-		if (parent == null) {
-			parent = currentTraceContext.get();
-		}
-		if (parent != null) {
-			return new ScopePassingSpanSubscriber<>(sub, context, currentTraceContext,
-					parent);
-		}
-		else {
-			return sub; // no need to trace
+	/**
+	 * Like {@link CurrentTraceContext#get()}, except it first checks the reactor context.
+	 */
+	// TODO: context is nullable likely for tests, but this is unverified
+	static TraceContext traceContext(@Nullable Context context,
```

Comment:
Will we ever expect `subscription.currentContext()` to return null? If not, I really want to clean this up..

Reply:
reactor-core uses `@NonNullApi` on a package level 
https://github.com/reactor/reactor-core/blob/master/reactor-core/src/main/java/reactor/core/package-info.java#L22-L23
``` 
@NonNullApi
package reactor.core;
```

and `CoreSubscriber#currentContext` does not have `@Nullable` annotation on return type 
https://github.com/reactor/reactor-core/blob/master/reactor-core/src/main/java/reactor/core/CoreSubscriber.java#L44  
```java
	default Context currentContext(){
		return Context.empty();
	}
```
So it must be safe assumption that context should not be null